### ISSUE TITLE
Use ragged arrays for alms

### ISSUE THREAD

**Issue by ntessore** [glass-dev/glass#487]
### Add your issue here

Inspired by s2fft, I think it would be great to use ragged `alm` arrays. The first axis would be $\ell = 0, \ldots, \ell_{\max}$, the second axis would be ragged with $m = 0, \ldots, \ell$.

**Comment by ntessore** [comment]
See also #382.

**Comment by Saransh-cpp** [comment]
Is s2fft already using ragged arrays from some package right now? I could not find anything relevant. Their paper says -

> From a software engineering perspective, this raises some further complications. Ideally, the FFTs of Eq. [(31)](https://www.sciencedirect.com/science/article/pii/S0021999124003589#fm0310) and Eq. [(32)](https://www.sciencedirect.com/science/article/pii/S0021999124003589#fm0320) would be implemented as a single function call to jax.numpy.fft across a regular array. However, jax.numpy.fft does not support ragged arrays and therefore each ring within the [polar region](https://www.sciencedirect.com/topics/earth-and-planetary-sciences/polar-region) must computed by individual function calls within a scheduled loop.

but I am not sure where the ragged array (that they want to feed in `jax.numpy.fft`) is coming from.

**Comment by ntessore** [comment]
You can see [here](https://astro-informatics.github.io/s2fft/api/sampling/spherical_samples.html#s2fft.sampling.s2_samples.flm_2d_to_hp) the format s2fft uses to store their alms:

```math
    \mathtt{alm}
    = \begin{bmatrix}
        0 & 0 & flm_{(0,0)} & 0 & 0 \\
        0 & flm_{(1,-1)} & flm_{(1,0)} & flm_{(1,1)} & 0 \\
        flm_{(2,-2)} & flm_{(2,-1)} & flm_{(2,0)} & flm_{(2,1)} & flm_{(2,2)}
    \end{bmatrix}
```

This is always for complex fields, so there are negative $m$ values present.

We would continue to use pairs of $E$ and $B$ mode alms instead, so no negative $m$ values. Our arrays could therefore look like this:

```math
    \mathtt{alm}
    = \begin{bmatrix}
        flm_{(0,0)} & 0 & 0 \\
        flm_{(1,0)} & flm_{(1,1)} & 0 \\
        flm_{(2,0)} & flm_{(2,1)} & flm_{(2,2)}
    \end{bmatrix}
```

Representing this as a ragged array makes it very easy to apply functions of $\ell$, e.g., when random sampling: just multiply over the leading axis. Selecting individual $m$ values would also very naturally do the right thing, e.g., `alm[:, 2:]` would have two empty rows, since there are no $m < 2$ values at $\ell = 0, 1$.

The only major issue here is that `ragged` doesn't support JAX 😞

**Comment by Saransh-cpp** [comment]
I see, thanks! I did notice this format, but the indexing looked very different from the one in GLASS, so I was skeptical.

`ragged` supports GPU and JAX through [Awkward Array](https://github.com/scikit-hep/awkward). [GPU support](https://awkward-array.org/doc/main/user-guide/how-to-use-in-numba-cuda.html) is extensive (including Numba + CUDA), but [JAX support](https://awkward-array.org/doc/main/user-guide/how-to-specialize-differentiate-jax.html) is limited (mostly due to restrictions from JAX). There are ongoing talks of writing a custom autodiff backend for Awkward and they might do it this summer.

I am not very sure how this is to be implemented but I am expecting that the alms should start looking like gls in the function arguments -
```py
gls = [gl_00,
       gl_11, gl_10,
       gl_22, gl_21, gl_20,
       ...]
```

but yes, it will become more clear once I have a quick meet with you after FOSDEM.
